core/src/utils/TTLCache.h

```diff
@@ -0,0 +1,86 @@
+/*
+ *
+ * TTLCache
+ *
+ * Created by El Khalil Bellakrid on 22/11/2019.
+ *
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2019 Ledger
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ */
+
+
+#ifndef LEDGER_CORE_TTLCACHE_H
+#define LEDGER_CORE_TTLCACHE_H
+
+#include <chrono>
+#include <unordered_map>
+#include <mutex>
+#include <iostream>
+#include "Option.hpp"
+/*
+ * A really simple implementation of TTL cache
+ */
+namespace ledger {
+    namespace core {
+        template <typename K, typename V, typename Duration = std::chrono::seconds>
+        class TTLCache {
+        public:
+            TTLCache(const Duration &ttl) : _ttl(ttl) {};
+
+            Option<V> get(const K &key) {
+                std::lock_guard<std::mutex> lock(_lock);
+                auto it = _cache.find(key);
+                if (it == _cache.end() ||
+                        std::chrono::duration_cast<Duration>(std::chrono::steady_clock::now().time_since_epoch()) - (*it).second.second > _ttl) {
+                    if (it != _cache.end()) {
```

Comment:
This is an open suggestion so I let you decide if you will apply it or not.

I am not a big fan to that double check on `it` so I suggest to to replace it by something like :
```cpp
if (it == _cache.end()) {
   return Option<V>();
}
else if (std::chrono::duration_cast<Duration>(std::chrono::steady_clock::now().time_since_epoch()) - (*it).second.second > _ttl) {
   _cache.erase(key);
  return Option<V>();
}
```

Even if the two `return` look redundant we can now return different values depending of the user case  (imagine an extra parameter telling us to valid the outdated cache for a new period of time).

Reply:
Hmm this is a good thought indeed, I was hesitating between redundant `return` or the double check on `it` but you are right, this can let us more freedom about how to make this code easily evolve